### PR TITLE
perslay

### PR DESCRIPTION
**Pull request by MathieuCarriere**
A pull request for adding the PersLay Tensorflow model in the representations module

### CODE REVIEW

File: src/python/gudhi/representations/perslay_params.md
Comment (VincentRouvreau) [inline src/python/gudhi/representations/perslay_params.md:103]: src/python/gudhi/representations/perslay_params.md should be moved in src/python/doc. Then you can do in src/python/doc/representations.rst :
```
.. include:: perslay_params.md
```
But here the render is not nice at all.
Either you rewrite it in restructured text, either we add the module https://github.com/miyakogi/m2r#sphinx-integration to the documentation generation process.
And then we could do :
```
.. mdinclude:: perslay_params.md
```
And the render shall be nice (I didn't try)

**Review by VincentRouvreau — COMMENTED** [review]
in src/python/doc/representations.rst, you need to add somewhere :
```
Perslay
-------
.. automodule:: gudhi.representations.perslay
   :members:
   :special-members:
   :show-inheritance:
```
This will show the PerslayModel documentation

**Comment by VincentRouvreau** [comment]
On https://6727-174304137-gh.circle-artifacts.com/0/htmlcov/gudhi_representations_perslay_py.html you can see that the code coverage of your tests are only 43%. I am pretty sure you could some tests for the non-tested layers. As Python code is not compiled, this would help to find regression on tensorflow versions for instance.

**Comment by MathieuCarriere** [comment]
OK sure I'll work on adding more tests

File: src/python/gudhi/representations/perslay_params.md
Comment (MathieuCarriere) [reply src/python/gudhi/representations/perslay_params.md]: Thanks for the tips! I improved the doc, it should look better now.

File: src/python/gudhi/representations/perslay.py
Comment (MathieuCarriere) [reply src/python/gudhi/representations/perslay.py:47]: That's actually a good question ;-) Roughly speaking, PersLay generalizes these methods. The functions "landscape_layer", "image_layer" etc can be used to compute these descriptors but also allow to compute new ones (that look similar, hence the name).

**Comment by mglisse** [comment]
Any idea what might cause this in the doc test?
```
Exception raised:
    Traceback (most recent call last):
      File "/usr/local/miniconda/envs/gudhi_build_env/lib/python3.6/doctest.py", line 1330, in __run
        compileflags, 1), test.globs)
      File "<doctest default[0]>", line 11, in <module>
        phi = gdr.GaussianPerslayPhi((100, 100), ((-.5, 1.5), (-.5, 1.5)), .1)
    AttributeError: module 'gudhi.representations' has no attribute 'GaussianPerslayPhi'

```

**Comment by MathieuCarriere** [comment]
Yep there was an error in the doc, hopefully the new push will fix it.

**Comment by MathieuCarriere** [comment]
@tlacombe feel free to have a look at the code

**Review by mglisse — COMMENTED** [review]
I had old comments that I apparently forgot to submit. I removed most of them because they seemed outdated, but let me send the remaining ones now to simplify things for me.

File: src/python/doc/representations.rst
Comment (mglisse) [inline src/python/doc/representations.rst:139]: We may want to specify the order in which the classes appear in the doc. By default, this is the lexicographic order. But it could make sense to group the Phi together, group the Weight separately, and have the main Perslay at one extremity.

File: src/python/gudhi/tensorflow/perslay.py
Comment (mglisse) [inline src/python/gudhi/tensorflow/perslay.py:243]: The links don't seem to work. Maybe ``:class:`~gudhi.tensorflow.perslay.GridPerslayWeight` `` would work better?

File: src/python/doc/representations.rst
Comment (mglisse) [inline src/python/doc/representations.rst:139]: Note that this way we end up documenting it as `gudhi.tensorflow.perslay.Perslay`. If you want to document it as `gudhi.tensorflow.Perslay`, we have to do it slightly differently I guess.

File: src/python/gudhi/representations/perslay_params.md
Comment (mglisse) [reply src/python/gudhi/representations/perslay_params.md]: There is a file `src/python/doc/perslay_params.md`, but I don't see it referenced anywhere :confused:

File: src/python/gudhi/tensorflow/perslay.py
Comment (mglisse) [inline src/python/gudhi/tensorflow/perslay.py:245]: Presumably it also works if I pass `perm_op = lambda x : tf.math.top_k(x, k=50)` (probably requires a few tweaks) instead of `perm_op = "top50"`? Hmm, not quite. I see the code calls `perm_op` with an argument `axis=1`, but the documentation does not tell me my custom function must support a parameter `axis`...

File: src/python/gudhi/tensorflow/perslay.py
Comment (mglisse) [inline src/python/gudhi/tensorflow/perslay.py:251]: I thought we had discussed replacing pop with perm_op so it is clearer?

File: src/python/gudhi/tensorflow/perslay.py
Comment (mglisse) [inline src/python/gudhi/tensorflow/perslay.py:244]: I guess I should look at the documentation of one of the existing Phi classes to find out what is expected of a custom function?

Comment (MathieuCarriere) [reply src/python/gudhi/tensorflow/perslay.py]: Yes, I added a comment about this in the doc.

File: src/python/gudhi/tensorflow/perslay.py
Comment (MathieuCarriere) [reply src/python/gudhi/tensorflow/perslay.py]: yep, just corrected this.

File: src/python/gudhi/tensorflow/perslay.py
Comment (MathieuCarriere) [reply src/python/gudhi/tensorflow/perslay.py]: I added a comment about this in the doc.

File: src/python/doc/representations.rst
Comment (MathieuCarriere) [reply src/python/doc/representations.rst:139]: Thanks, I've updated it and it works now.

File: src/python/gudhi/tensorflow/perslay.py
Comment (MathieuCarriere) [reply src/python/gudhi/tensorflow/perslay.py]: Thanks, I've updated it and it works now.

File: src/python/gudhi/tensorflow/perslay.py
Comment (MathieuCarriere) [reply src/python/gudhi/tensorflow/perslay.py]: Yep, just removed the import.

File: src/python/doc/installation.rst
Comment (MathieuCarriere) [reply src/python/doc/installation.rst]: Thanks, I've just fixed it.

**Comment by mglisse** [comment]
As can be seen by clicking on the CI results at the bottom, the doctest (representations.rst or vector_methods.py.rst) fails
```
Expected:
    tf.Tensor(
    [[[[1.7266072e-16]
...
       [7.8093526e-16]]]], shape=(1, 5, 5, 1), dtype=float32)
Got:
    tf.Tensor(
    [[[[1.7266072e-16]
...
       [0.0000000e+00]]]], shape=(1, 100, 100, 1), dtype=float32)
```

**Comment by mglisse** [comment]
Do you mind if I push https://github.com/mglisse/gudhi-devel/commit/224b13d5884598c5d4613ca2c8fc321e39153165 to your branch? (then you will need a pull on your local clone to get everything back in order)

**Comment by MathieuCarriere** [comment]
sure, go ahead

**Review by mglisse — APPROVED** [review]
I still think it would be cleaner not to import things in gudhi.tensorflow (`import gudhi.tensorflow.perslay as prsl` and then using `prsl.GridPerslayWeight` does not seem less convenient than what you have now), but I think I lost this battle a long time ago.
I also believe the representations doc page was already too long before perslay and should be split. It isn't even clear that perslay belongs there, with its non-sklearn interface. But that doesn't block the PR for me, we can improve that later.

**Comment by MathieuCarriere** [comment]
Sorry for being so slow on this PR. I am not against your suggestions, I think they just slipped off my mind. Same thing with the doc, for which I also agree with you. My goal is to implement these changes this week.

**Comment by mglisse** [comment]
As I was saying, the doc change is not urgent, we can still do it after the merge in a separate PR.
For the import thing, I thought you were against it... If you are not, I can quickly do that change for you.

**Comment by mglisse** [comment]
> For the import thing, I thought you were against it... If you are not, I can quickly do that change for you.

There, that should work, although you may want to rename prsl to something nicer. If that's not what you meant, feel free to revert this change, after all, I approved the PR before that :wink:

**Comment by MathieuCarriere** [comment]
Great thank you! I had a look at it again, but cannot remember why I was against it... I'm perfectly fine with it anyway!

**Pull request merged**
